drafts/LET-DRAFT.md

```diff
@@ -0,0 +1,577 @@
+    for each e in LET SOURCE:
+        let_binding = let_binding + ⟨r_current:e_current⟩
+    add (b + let_binding) to Cur
+    Prev = Current
+    Current = <<>>
+return Prev
+```
+
+Notice that
+
+```
+`LET 1     AS x 
+    2     AS y 
+    x + y AS x `
+```
+
+will result in:
+
+```
+`x = 1
+y = 2 
+x = 3 
+`
+```
+
+#### **Definition - Option 3**
+
+This definition uses the concatenate operator `||` when dealing with `LET` source, and uses the newly defined `+` operator to add **bind names** to the current **variable environment**.
+
+```
+Prev <- Bin_Let
+Cur <- <<>>  // an empty bag
+for each binding tuple b in Prev:
+    let_binding = ⟨⟩
+    for each e in LET SOURCE:
+        let_binding = let_binding || ⟨r_current:e_current⟩
+    add (b + let_binding) to Cur
+    Prev = Current
+    Current = <<>>
+return Prev
+```
+
+This way:  Notice that
+
+```
+`LET 1     AS x 
+    2     AS y 
+    x + y AS x `
+```
+
+will remain as:
+
+```
+`y = 2 
+x = 3 `
+```
+
+
+The implication of each option will be further demonstrated in the next section. To reduce the length, we will mostly compare Option 1 with Option 3.
+* * *
+
+### **LET as an optional following clause to any other clause**
+
+With the proposed `LET` definition, let us explore the possibility of extends the support for `LET` to other clauses.
+
+```
+FROM ...
+LET ...
+WHERE ...
+LET ...
+GROUP BY ...
+LET ...
+HAVING ...
+LET ...
+ORDER BY ...
+LET ...
+OFFSET ...
+LET ...
+LIMIT ...
+LET ...
+SELECT ...
+LET ...
+```
+
+Consider the following to example:
+
+```
+FROM a LET b AS x, c AS y WHERE x > 1
+```
+
+```
+FROM a LET b AS x WHERE x > 1 LET c AS y
+```
+
+Now notice the **the bag of binding tuples** produced by the two queries are exact the same. One could argue that a query optimizer should be able to decide where the `LET` should be executed.
+
+With this in mind, the question we should focus on first is, should we support `LET` after `GROUP BY`, `LET` after `ORDER BY` and `LET` after `SELECT`.
+* * *
+
+#### **GROUP BY/ ORDER BY**
+
+Notice that since the proposed `LET` semantics follows the pattern of “**input bag of binding tuples, evaluate constituent expressions, output bag of binding tuples**”, it should have no problem following `GROUP BY` and `ORDER BY`.
+For example:
+
+```
+From log as l GROUP BY l.sensor AS sensor GROUP AS g LET 1 as x
+```
+
+Suppose that:
+
+```
+Bout_GROUPBY = Bin_Let 
+         = **<<**
+            ⟨sensor:1, 
+             g: << ⟨l:{’sensor’:1, ’co’:0.4}⟩,⟨l:{’sensor’:1, ’co’:0.2}>>
+            ⟩,
+            ⟨sensor:2, 
+             g:<< ⟨l:{’sensor’:2, ’co’:0.3}⟩ >>**
+            ⟩ 
+           >> 
+**
+```
+
+Then:
+
+```
+Bout_Let = <<
+            ⟨sensor:1, 
+             g: << ⟨l:{’sensor’:1, ’co’:0.4}⟩,⟨l:{’sensor’:1, ’co’:0.2}>>,
+             x:1
+            ⟩,
+            ⟨sensor:2, 
+             g:<< ⟨l:{’sensor’:2, ’co’:0.3}⟩ >>,
+             x:1
+            ⟩ 
+           >> 
+```
+
+* * *
+
```

Comment:
Added To-do notes. Not immediately resolved because: 
`ORDER BY` essentially is is just converting the bag of binding tuple to a list of binding tuple, LET behavior after ORDER BY is relatively predictable.